src/cloupe/cloupe.py

```diff
@@ -0,0 +1,251 @@
+        self.celltracks = []
+        if self.header["nextHeaderOffset"] != self.get_file_size():
+            self.next_header = self.read_header(self.header["nextHeaderOffset"])
+            self.next_index_block = self.read_block(
+                start=self.next_header["indexBlock"]["Start"],
+                end=self.next_header["indexBlock"]["End"],
+                as_json=True,
+            )
+            # now that we have the index block we get the 'userCreated' celltracks
+            self.celltracks = []
+            for celltrack in self.next_index_block["CellTracks"]:
+                celltrack["Metadata"] = self.read_block(
+                    start=celltrack["Metadata"]["Start"],
+                    end=celltrack["Metadata"]["End"],
+                    as_json=True,
+                )
+                vblock_bytes = self.read_block(
+                    start=celltrack["Values"]["Start"],
+                    end=celltrack["Values"]["End"],
+                )
+                vblock_count = celltrack["Values"]["ArraySize"]
+                groups = celltrack["Metadata"]["groups"]
+                group_keys = struct.unpack(f"{vblock_count}h", vblock_bytes)
+                celltrack["Values"] = [groups[gk] if gk!=-1 else "UNASSIGNED" for gk in group_keys]
+                self.celltracks.append(celltrack)
+
+    def read_header(self, position=0):
+        with open(self.cloupe_path, "rb") as f:
+            f.seek(position)
+            header = f.read(4096)
+            return json.loads(header.decode("utf-8", errors="replace").strip("\x00"))
+
+    def get_file_size(self):
+        with open(self.cloupe_path, "rb") as f:
+            f.seek(0, 2)
+            size = f.tell()
+            return size
+
+    def __repr__(self):
+        return str(self.index_block["Runs"])
+
+    # trick for partial object decompression
+    def decompress(self, chunk):
+        gz = zlib.decompressobj(31)
+        return gz.decompress(chunk)
+
+    # read cloupe block
+    def read_block(self, start, end, as_json=False):
+        with open(self.cloupe_path, "rb") as f:
+            f.seek(start)
+            read_size = end - start
+            block = f.read(read_size)
+            if block[:2] == b"\x1f\x8b":
+                logging.debug("compressed block")
+                block = self.decompress(block)
+            if as_json:
+                return json.loads(block.decode("utf-8", errors="replace"))
+            return block
+
+    # writing the barcodes to the csv file
+    def barcodes_writer(self):
+        barcodes = self.cwd + '/barcodes.csv'
+        with open(barcodes, 'w', newline="") as barcodes_file:
+            # Initialize the CSV writer
+            csv_writer = csv.writer(barcodes_file)
+            csv_writer.writerow(["Barcodes"])
+            csv_writer.writerows(
+                    [[element] for element in self.matrices[0]['Barcodes'][:self.matrices[0]['BarcodeCount']]]
```

Comment:
What's the purpose of limiting this to the length of `self.matrices[0]['BarcodeCount']]` ?

Reply:
Done